### PR TITLE
WebAssembly 2021 chapter

### PR DESCRIPTION
**Pull request by RReverser**
Makes progress on #2168 

Staging URL: https://20211118t191759-dot-webalmanac.uk.r.appspot.com/en/2021/webassembly

### CODE REVIEW

**Comment by tunetheweb** [comment]
Turned it on in config and staged it here: https://20211118t191759-dot-webalmanac.uk.r.appspot.com/en/2021/webassembly (link added to first comment). @rviscomi or I can update this URL with the latest periodically.

**Comment by rviscomi** [comment]
@RReverser is also running the site locally so it's less critical that we stage it for the author's benefit.

**Comment by tunetheweb** [comment]
Ah Ok wasn’t sure if he was as config wasn’t set. Plus was nosy myself to have a look 😁

**Comment by RReverser** [comment]
> Ah Ok wasn’t sure if he was as config wasn’t set. Plus was nosy myself to have a look 😁

Ah thanks, yeah I didn't commit the config change as I was unsure if I'm supposed to leave that to maintainers :)

**Comment by tunetheweb** [comment]
Yeah will flip it back before merging so it won't accidentally go live before we give it a round of editing. But find it best to have it on for this stage for convenience without having to remember not to commit it each time!

**Comment by RReverser** [comment]
Still need to address some TODOs in metadata, but content-wise this should be ready for the first round of feedback.

**Comment by RReverser** [comment]
To clarify - I mean, editorial review, as we are done with peer review in Google Docs.

File: src/content/en/2021/webassembly.md
Comment (tunetheweb) [inline src/content/en/2021/webassembly.md:346]: I struggle how to apply the aforementioned chart changes to this graph since it shows so little EXCEPT for the extreme percentiles (0, and in particular 100).

So I rerun the stats using all 100 percentiles (using `UNNEST(GENERATE_ARRAY(0, 100)) AS percentile)) to see what it looked like, and put the results here: https://docs.google.com/spreadsheets/d/1IMa2SbdQgshb4pGWF1KOh9s4zMtLbRymWZGYjdaatXY/edit#gid=50034228

The produces this graph:

image

You can see there really is nothing until right at the end, when it's a handful of (badly optimised?) examples.

Here's the raw data:

percentile | desktop | mobile
-- | -- | --
0 | -501 | -501
1 | 0 | 0
2 | 0 | 0
... | ... | ...
87 | 0 | 0
88 | 0 | 0
89 | 0 | 16
90 | 22 | 89
91 | 103 | 116
92 | 122 | 1,030
93 | 1,528 | 3,397
94 | 20,103 | 42,470
95 | 55,084 | 55,084
96 | 96,876 | 93,179
97 | 120,361 | 120,358
98 | 120,361 | 120,361
99 | 252,733 | 133,527
100 | 2,486,838 | 1,291,316


Should we just not have a graph and instead have a Big Number and/or discussion?
Or a table
Does this extra information change anything you want to say here?

File: src/content/en/2021/webassembly.md
Comment (tunetheweb) [inline src/content/en/2021/webassembly.md:401]: Similar comment to above. Though think you've explained it better in the text here.

image

percentile | desktop | mobile
-- | -- | --
0 | -260,632 | -178,826
1 | -118,704 | -66,562
2 | -118,704 | -49,206
3 | -118,704 | -39,136
4 | -118,704 | -27,119
5 | -85,564 | -20,875
6 | -50,169 | -16,809
7 | -43,979 | -16,236
8 | -29,499 | -10,592
9 | -27,119 | -5,987
10 | -27,119 | -5,890
11 | -25,982 | -5,890
12 | -16,809 | -5,890
13 | -14,181 | -5,890
14 | -9,194 | -5,890
15 | -5,890 | -5,890
16 | -5,890 | -5,890
17 | -5,890 | -5,890
18 | -5,890 | -5,890
19 | -5,890 | -5,890
20 | -5,890 | -5,890
21 | -5,890 | -5,890
22 | -5,890 | -5,890
23 | -5,890 | -5,890
24 | -5,890 | -5,890
25 | -5,890 | -5,775
26 | -5,890 | -5,775
27 | -5,890 | -5,775
28 | -5,890 | -5,775
29 | -5,890 | -5,775
30 | -5,890 | -5,775
31 | -5,890 | -5,775
32 | -5,890 | -5,775
33 | -5,890 | -5,775
34 | -5,890 | -5,775
35 | -5,890 | -5,775
36 | -5,890 | -5,775
37 | -5,890 | -5,656
38 | -5,775 | -4,850
39 | -5,775 | -4,788
40 | -5,775 | -4,788
41 | -5,775 | -4,305
42 | -5,775 | -4,142
43 | -5,775 | -3,647
44 | -5,775 | -2,792
45 | -4,802 | -2,419
46 | -4,788 | -2,419
47 | -4,456 | -2,306
48 | -4,207 | -1,754
49 | -3,877 | -1,117
50 | -3,647 | -611
51 | -2,792 | -492
52 | -2,419 | -268
53 | -2,306 | -220
54 | -1,945 | -119
55 | -1,167 | -119
56 | -1,117 | -38
57 | -569 | -12
58 | -492 | -12
59 | -248 | -12
60 | -119 | -12
61 | -119 | -12
62 | -45 | -12
63 | -12 | -12
64 | -12 | -11
65 | -12 | -7
66 | -12 | -7
67 | -12 | -1
68 | -12 | 0
69 | -10 | 0
70 | -7 | 0
71 | 0 | 0
72 | 0 | 2
73 | 0 | 3
74 | 0 | 5
75 | 2 | 6
76 | 4 | 7
77 | 5 | 10
78 | 10 | 16
79 | 16 | 22
80 | 22 | 22
81 | 28 | 30
82 | 30 | 30
83 | 32 | 32
84 | 32 | 32
85 | 32 | 32
86 | 34 | 32
87 | 77 | 32
88 | 187 | 32
89 | 314 | 34
90 | 513 | 77
91 | 730 | 246
92 | 1,440 | 443
93 | 2,626 | 600
94 | 5,556 | 1,039
95 | 7,171 | 3,135
96 | 11,910 | 5,556
97 | 12,938 | 7,171
98 | 24,465 | 11,316
99 | 24,465 | 12,938
100 | 1,032,341 | 246,911

**Review by tunetheweb — CHANGES_REQUESTED** [review]
I redid some of the images to:
- Use standard font size of 16px.
- Hide green label when it didn't fit,
- Avoid black label on green data points (doesn't look as nice).
- Hide 0 and 100 percentile. These are, by definition, extremes and while it's fine to talk about them in the text (they often make interesting data points!) we generally avoid putting them on the graphs as they tend to skew the impression with these extremes.
- Stop using logarithmic scale (less necessary with the above).
- Move from bytes to kilobytes to make labels easier to read.
- Move to horizontal graphs when it fits (particularly for percentiles as thats what we use in the Web Almanac).
- Retake images with these changes and TinyPng them

You can compare the results:
- Original: https://20211118t191759-dot-webalmanac.uk.r.appspot.com/en/2021/webassembly?print
- Latest: https://20211121t105115-dot-webalmanac.uk.r.appspot.com/en/2021/webassembly?print

(note I'm using the `?print` query param to show the images as Sheets are changing)

Hopefully you'll agree to these changes.

I've two more graphs to do but have questions on them, highlighted below.

After this we'll just need a quick copy edit of the text and should be done.

**Comment by RReverser** [comment]
> * Use standard font size of 16px.
> * Hide green label when it didn't fit,

I actually wanted to show both - that's why I reduced font size or increased chart height in few places, as otherwise it's hard to reason about differences between mobile and desktop looking at the graph. On desktop readers can at least hover over labels (although that's a bit cumbersome), but on mobile they don't have even that.

> * Avoid black label on green data points (doesn't look as nice).

I thought we specifically avoided green labels because they're not very accessible (low-contrast). Dataviz templates currently reflect your comment [here](https://github.com/HTTPArchive/almanac.httparchive.org/issues/2217#issuecomment-901982315) and use black labels for green graphs - that's why I copied and went with the same colours too. I'd rather keep the consistency and higher contrast, although I agree it's not as pretty. 

> * Hide 0 and 100 percentile. These are, by definition, extremes and while it's fine to talk about them in the text (they often make interesting data points!) we generally avoid putting them on the graphs as they tend to skew the impression with these extremes.

It's something I've added later in the process. Initially I started with standard 10-90 range, but in this dataset extremes are particularly interested - as you noticed, in some graphs _only_ extremes have any data at all, which is an interesting datapoint by itself.  In some graphs `0` extreme even reveals regressions which are a very interesting point by itself.

I agree it's fine to talk about them in text, but visualization helps to understand them better IMO, and minimum/maximum sizes and savings are particularly interesting.

> * Stop using logarithmic scale (less necessary with the above).
> * Move from bytes to kilobytes to make labels easier to read.
> * Move to horizontal graphs when it fits (particularly for percentiles as thats what we use in the Web Almanac).
> * Retake images with these changes and TinyPng them

All of this sounds good.

File: src/content/en/2021/webassembly.md
Comment (RReverser) [reply src/content/en/2021/webassembly.md:346]: > Does this extra information change anything you want to say here?

Not really, that's part of what I wanted to show - that most toolchains already strip those sections, however in the few that don't, manually stripping custom sections can save even more.

Comment (RReverser) [reply src/content/en/2021/webassembly.md:346]: > I was considering making a graph over only unstripped modules

Maybe it would help to add such graph as an extra here though, and explain why I'm doing it? That could work better, because I agree that the extreme is a bit different / misleading in this case.

**Comment by tunetheweb** [comment]
> > * Hide green label when it didn't fit,
> 
> I actually wanted to show both - that's why I reduced font size or increased chart height in few places, as otherwise it's hard to reason about differences between mobile and desktop looking at the graph. On desktop readers can at least hover over labels (although that's a bit cumbersome), but on mobile they don't have even that.
> 
> > * Avoid black label on green data points (doesn't look as nice).
> 
> I thought we specifically avoided green labels because they're not very accessible (low-contrast). Dataviz templates currently reflect your comment [here](https://github.com/HTTPArchive/almanac.httparchive.org/issues/2217#issuecomment-901982315) and use black labels for green graphs - that's why I copied and went with the same colours too. I'd rather keep the consistency and higher contrast, although I agree it's not as pretty.

It's something we've gone back and forth ourselves quote a bit as you've notice. There's no great answers here. And I almost look for excuses to drop the data label to avoid the issue 😁

But taking the first graph as an example I find the [revised one](https://20211121t105115-dot-webalmanac.uk.r.appspot.com/en/2021/webassembly?print#fig-1) a lot easier to take in than the [original one](https://20211118t191759-dot-webalmanac.uk.r.appspot.com/en/2021/webassembly?print#fig-1) — especially on mobile!

image

To allow both we need to drop the font size to a less readable size AND it still looks like a squeeze anyway (not to mention the black/green issue). Dropping one of the labels is used a lot across the Web Almanac for these reasons. The size of desktop can be inferred approximately from mobile label and we have the description (we can update this to include both stats) **and** access to the underlying sheet from the 3 dots menu should readers want the real data.

> > * Hide 0 and 100 percentile. These are, by definition, extremes and while it's fine to talk about them in the text (they often make interesting data points!) we generally avoid putting them on the graphs as they tend to skew the impression with these extremes.
> 
> It's something I've added later in the process. Initially I started with standard 10-90 range, but in this dataset extremes are particularly interested - as you noticed, in some graphs _only_ extremes have any data at all, which is an interesting datapoint by itself. In some graphs `0` extreme even reveals regressions which are a very interesting point by itself.
> 
> I agree it's fine to talk about them in text, but visualization helps to understand them better IMO, and minimum/maximum sizes and savings are particularly interesting.

Maybe we just need to think of better visualizations here then? I found the use of logarithmic scale quite confusing (especially as it's not something we use a lot (any?) of in the Web Almanac and it wasn't always apparent this scale was used unless you really scrutinised the numbers) — if the middle numbers aren't what we care about then no need to use that to allow the middle numbers to be seen. Maybe the 100 percentile point graph above is better to make your point? Or a table of just the extremes? Or a separate visualisation(s) looking only at the extreme ends?

However, we also need to be careful as to what data we can draw from the extreme. For example a web performance chapter wouldn't necessary spend a huge amount of time talking about sites that are > 40MB, or sites that score 0 on Lighthouse because those are (thankfully!) rare. The 100th percentile in particular is the absolute worst case and a single bad site will massively throw that out. There's always going to be outliers and if the point you're making is only relevant to a very, very small number, or extreme sites, then while it might be something to call out as a warning, I'm not sure it's good to spend a several visualisations and sections on it. But you're the author so will leave you to make that call. I'm just trying to ensure readability and consistency across the Web Almanac. 😁

**Comment by RReverser** [comment]
> Dropping one of the labels is used a lot across the Web Almanac for these reasons.

Sure, for something like mime types one dataset is enough and they're pretty close anyway.

In the section with size graphs, however, the differences between desktop and mobile are much more pronounced, so I was really hoping to show numbers for both.

> if the middle numbers aren't what we care about then no need to use that to allow the middle numbers to be seen

I think we do care about both middle numbers (especially 10, 50 and 90) as well as the extremes. That's why percentile graph still seemed most appropriate, even if I added two new data points.

I agree the logarithmic scale is best avoided, but not sure how else to show all those data points visually.

> For example a web performance chapter wouldn't necessary spend a huge amount of time talking about sites that are > 40MB

Ah hmm I thought other sections would call them out too as those are pretty interesting extremes. In our case, we also went with "what if we optimized those Wasms" which shows how easy it is to fix those huge Wasms with few extra steps.

> Or a separate visualisation(s) looking only at the extreme ends?

I like this alternative most so far, at least if it's the only way to avoid log scale. It might get a bit brittle repeating two visualizations for each step but maybe not too bad.

**Comment by tunetheweb** [comment]
> In the section with size graphs, however, the differences between desktop and mobile are much more pronounced, so I was really hoping to show numbers for both.

Oh I think you may have misunderstood me here, I didn't drop them in ALL cases — I left them in for those graphs as there was the space:

image

I did change them to green labels though, as that's what we've gone with elsewhere. Really don't like the black to be honest - works fine on inner green labels but really doesn't on outer green labels. Plus the description covers us a little from an a11y point. We also added a note to [our Accessibility Statement](https://github.com/HTTPArchive/almanac.httparchive.org/pull/2511) (not that that excuses us, but at least owning up to the issue!). It's not ideal, I admit, but unfortunately that is the Web Almanac style for now.

**Comment by tunetheweb** [comment]
We could go back to logarithmic scale foe some of these if you think it's better? Just think we need to be pretty clear it's that.

**Comment by RReverser** [comment]
> Oh I think you may have misunderstood me here, I didn't drop them in ALL cases — I left them in for those graphs as there was the space:

Ah right. I think it's worth returning them back in this graph too, or splitting it into two graphs if it doesn't fit: https://20211121t105115-dot-webalmanac.uk.r.appspot.com/static/images/2021/webassembly/popular_by_name.png

Otherwise right now we have few libraries that are present only on mobile yet don't have labels, as well as few where usage significantly differs.

> It's not ideal, I admit, but unfortunately that is the Web Almanac style for now.

Yeah I wish there was a better solution, but leaving that up to you :) Meanwhile it might be worth updating the dataviz templates if that's indeed the style we go with for all articles.

**Comment by RReverser** [comment]
> We could go back to logarithmic scale foe some of these if you think it's better? Just think we need to be pretty clear it's that.

In general, it doesn't look as bad as I feared. I'm still torn on what to do about extremes though - showing them as part of the same graph still feels more natural, but you're also right that they do change the scales significantly and were pretty much the only reason to use log scale...

**Comment by rviscomi** [comment]
A table would be a good option if the extreme percentiles are needed and orders of magnitude different.

**Comment by RReverser** [comment]
> A table would be a good option if the extreme percentiles are needed and orders of magnitude different.

Do you mean a table in addition to the graph or instead of one?

**Comment by rviscomi** [comment]
I was thinking instead of a chart.

**Comment by tunetheweb** [comment]
> Ah right. I think it's worth returning them back in this graph too, or splitting it into two graphs if it doesn't fit: https://20211121t105115-dot-webalmanac.uk.r.appspot.com/static/images/2021/webassembly/popular_by_name.png
> 
> Otherwise right now we have few libraries that are present only on mobile yet don't have labels, as well as few where usage significantly differs.

Added back in in https://github.com/HTTPArchive/almanac.httparchive.org/pull/2563/commits/3873de29d6adae1f0665c2e552881566f3046932. Made the labels 14px rather than 16px and extended the height a bit more. Updated the staging link (might take a few mins for cache to update though).

> > It's not ideal, I admit, but unfortunately that is the Web Almanac style for now.
> 
> Yeah I wish there was a better solution, but leaving that up to you :) Meanwhile it might be worth updating the dataviz templates if that's indeed the style we go with for all articles.

Well we do use the black on inner green labels. Plus it stands out so one I always spot when doing a review. I wish sheets allowed a text outline. Green text with black outline would look OK I think. But alas it doesn't support that.

> > > A table would be a good option if the extreme percentiles are needed and orders of magnitude different.
> > 
> > Do you mean a table in addition to the graph or instead of one?
>
> I was thinking instead of a chart.

Do you wanna have a think about it @RReverser and update this branch? I'll not touch it any more until I hear back from you.

**Comment by RReverser** [comment]
> Do you wanna have a think about it @RReverser and update this branch? I'll not touch it any more until I hear back from you.

I think I'd like to keep the charts at least for 10-90 range (visualizations are so pretty! 😅) but yeah let me get back to you tomorrow.

**Comment by RReverser** [comment]
Is there anything to change in text or is someone else going to review it too?

**Comment by tunetheweb** [comment]
> Is there anything to change in text or is someone else going to review it too?

We'll do a copy edit of the text after this is merged. I like to get the controversial things out of the way in the main PR as much as possible, so that copyedit PR is easier for you to review — rather than everything having changed cause I rearranged the figures completely or the like 😁 Hopefully should just be a quick spelling and grammar check.

**Comment by tunetheweb** [comment]
> I think I'd like to keep the charts at least for 10-90 range (visualizations are so pretty! 😅)

+ 1 to this!

**Comment by RReverser** [comment]
Looks like this graph got cut off after the last update: 
image

**Comment by RReverser** [comment]
Ok I think going with just big numbers or text where appropriate is good enough after all. I've adjusted wording under a graph where we removed 0% percentile.

File: src/content/en/2021/webassembly.md
Comment (RReverser) [reply src/content/en/2021/webassembly.md:403]: I think this one can be converted into 10-90 range - it should look fine without log scale - and then I'll mention the extremes in text. Can you convert the graph into the same style as you did others please?

File: src/content/en/2021/webassembly.md
Comment (RReverser) [reply src/content/en/2021/webassembly.md:348]: Ok, please convert this graph into the same style. I'll then use it to show that most files already don't have custom sections, and I will add another graph right underneath to show savings over only files that did have those sections.

Comment (tunetheweb) [reply src/content/en/2021/webassembly.md:348]: Done, committed and staging update: https://20211121t105115-dot-webalmanac.uk.r.appspot.com/en/2021/webassembly#fig-18

Though not showing much when converted to kilobytes:

image

Should we revert to bytes for this one? Though that feels a little bit of a con just to inflate the numbers... 😀

image

File: src/content/en/2021/webassembly.md
Comment (tunetheweb) [reply src/content/en/2021/webassembly.md:403]: Done, committed and staging update: https://20211121t105115-dot-webalmanac.uk.r.appspot.com/en/2021/webassembly#fig-18

Though not showing much when converted to kilobytes:

image

Should we revert to bytes for this one? Though that feels a little bit of a con just to inflate the numbers... 😀

image

File: src/content/en/2021/webassembly.md
Comment (tunetheweb) [reply src/content/en/2021/webassembly.md:348]: Done: https://20211121t105115-dot-webalmanac.uk.r.appspot.com/en/2021/webassembly#fig-15

Pretty boring graph though:

image

File: src/content/en/2021/webassembly.md
Comment (RReverser) [reply src/content/en/2021/webassembly.md:403]: I guess we could keep KB but just add 2 (?) digits after a dot as a compromise between consistency (KB in all graphs) and usefulness.

File: src/content/en/2021/webassembly.md
Comment (RReverser) [reply src/content/en/2021/webassembly.md:348]: That's okay, as I mentioned, it's still useful to show the zeroes in this case and I'll add another graph below :)

Looking at this and the other graph though, I'm less sure that converting to KB was a good idea now... Aside from initial sizes, all differences are not that big and might be actually easier to read in bytes.

File: src/content/en/2021/webassembly.md
Comment (RReverser) [reply src/content/en/2021/webassembly.md:403]: Also see my comment above - should we maybe just revert to bytes for those? It makes it clearer when e.g. zero is actually zero and not a rounding error.

Comment (tunetheweb) [reply src/content/en/2021/webassembly.md:403]: Good idea! Also had my rang wrong so missed of the 10th percentile and messed up my desktop and mobile labels:

image

File: src/content/en/2021/webassembly.md
Comment (tunetheweb) [reply src/content/en/2021/webassembly.md:348]: Here it is in bytes:

image

Though to be honest I'm not sure 22 or even 89 bytes saving is much to shout about! 😀

File: src/content/en/2021/webassembly.md
Comment (tunetheweb) [reply src/content/en/2021/webassembly.md:403]: On this one I do think converting to bytes makes it more difficult to read:

image

But happy to leave it in bytes if you want?

Comment (RReverser) [reply src/content/en/2021/webassembly.md:403]: Hmm yeah overlapping labels are definitely not great. I guess we could change font size, but that only makes things worse.

The 2 digits after dot looks good enough for a compromise, let's leave that.

File: src/content/en/2021/webassembly.md
Comment (RReverser) [reply src/content/en/2021/webassembly.md:348]: You know what, let's just add your 100 percentile scatter plot here instead, and I'll use it as a basis to explain the proper percentile graph over files with custom sections.

Comment (tunetheweb) [reply src/content/en/2021/webassembly.md:348]: On the plus side, it's another graph type - variety is the spice of live! 🌶

Here's the details:
- chart_url: https://docs.google.com/spreadsheets/d/e/2PACX-1vT6yhkn3lw148YQQHLoqA71NIsZLSSoBtgFmd_hRyhcmyPl2OpLyuOjUBk64I5DLE_grN8esL8oA3zt/pubchart?oid=852876683&format=interactive
- sheets_gid: 50034228
- sql_file: sizes_and_savings_100.sql

Will update the image and add the SQL before merging.

Comment (RReverser) [reply src/content/en/2021/webassembly.md:348]: I've gathered percentiles over files with custom sections and added a new graph meanwhile:
image

File: src/content/en/2021/webassembly.md
Comment (RReverser) [reply src/content/en/2021/webassembly.md:403]: Marking this one as resolved as it's getting hard to track conversations 😅

File: src/content/en/2021/webassembly.md
Comment (RReverser) [reply src/content/en/2021/webassembly.md:12]: I added some featured stats but no idea what to pick for the quote 😅

Comment (RReverser) [reply src/content/en/2021/webassembly.md:12]: I guess I could copy this from the conclusion if it's not too long:

> While WebAssembly is a relatively new and somewhat niche participant on the web, it's great to see its adoption across a variety of websites and use-cases, from simple libraries to large applications. In fact, we could see that it integrates so well into the web ecosystem, that many website owners might not even know they already use WebAssembly—to them it looks like any other 3rd-party JavaScript dependency.

Comment (RReverser) [reply src/content/en/2021/webassembly.md:12]: I could reword it to be shorter here, but I guess quote must be an exact quote?

**Comment by RReverser** [comment]
FWIW I think I'm done with the graphs, please take another look and rerender where necessary.

File: src/content/en/2021/webassembly.md
Comment (tunetheweb) [reply src/content/en/2021/webassembly.md:12]: Can be select in your quote with the use of "...",

> WebAssembly... integrates so well into the web ecosystem, that many website owners might not even know they already use WebAssembly—to them it looks like any other 3rd-party JavaScript dependency.

or pick first part:

> While WebAssembly is a relatively new and somewhat niche participant on the web, it's great to see its adoption across a variety of websites and use-cases, from simple libraries to large applications.

Could also pick from introduction for someone not sure what WebAssembly is:

> WebAssembly is a binary instruction format that allows developers to compile code written in languages other than JavaScript and bring it to the web in an efficient, portable package. The existing use-cases range from reusable libraries and codecs to full GUI applications.

Normally would advise picking something a bit more interesting that just a definition, but given WebAssembly is new and niche maybe a definition isn't a bad thing here?

BTW if you run `npm run generate` locally and then look at: http://127.0.0.1/en/2021/?feat=webassembly you'll see what your featured stat/quote looks like.

Comment (RReverser) [reply src/content/en/2021/webassembly.md:12]: I went with `WebAssembly... integrates so well into the web ecosystem, that many website owners might not even know they already use WebAssembly—to them it looks like any other 3rd-party JavaScript dependency.`.

**Comment by tunetheweb** [comment]
Images updated. This good to merge now @RReverser ?

**Comment by RReverser** [comment]
I think so, thanks! I could definitely use some extra eyes on text and perhaps on KB -> MB conversions (I've noticed few places where I messed it up during last go through the text) but you said we'll do text review in a separate PR?

**Comment by tunetheweb** [comment]
Yup. Let's merge this and do that separate.

Thanks very much for being so patient with all my requests. but think we made it better so hopefully you'll agree it was worth it!

**Comment by RReverser** [comment]
> Yup. Let's merge this and do that separate.
> 
> Thanks very much for being so patient with all my requests. but think we made it better so hopefully you'll agree it was worth it!

I mean, I still miss my extremes, but I agree it looks more consistent and better overall 😀

**Pull request merged**
